### ISSUE TITLE
How does the process of YOLOv8 predict the class

### ISSUE THREAD

**Issue by DonyMimin** [ultralytics/ultralytics#16325]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

hi @glenn-jocher, I'm currently working on my research using yolov8. I wonder how does the process of YOLOv8 predict the classes? did it use fully connected on the head?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DonyMimin, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com) where you can find detailed explanations about YOLOv8's architecture and processing. You might also find our [Model](https://docs.ultralytics.com/models/) page useful for understanding how class prediction is handled.

If this is a 🐛 Bug Report, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) so we can further investigate.

If this is a custom training ❓ Question, please share as much detail as possible, including dataset examples and training logs, and confirm you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community for more interactive discussions! For real-time chat, check out [Discord](https://ultralytics.com/discord) 🎧. For more in-depth questions, visit [Discourse](https://community.ultralytics.com) or engage with us on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're using the latest version of the `ultralytics` package by upgrading in a [**Python>=3.8**](https://www.python.org/) environment with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed:

```bash
pip install -U ultralytics
```

## Environments

Run YOLOv8 in any verified environment with all dependencies like [CUDA](https://developer.nvidia.com/cuda), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled:

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)

<a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates current CI test status. Green means tests are passing, which ensures proper operation across different systems.

🔔 This is an automated response. An Ultralytics engineer will assist you soon!

**Comment by glenn-jocher** [comment]
@DonyMimin no, YOLOv8 uses only convolutional layers in the head and does not use fully connected or "linear" layers, other than in the Classify models. Detect() heads are here:
https://github.com/ultralytics/ultralytics/blob/c5bee7545cb5fe26e201c593f3920b4b462c73c1/ultralytics/nn/modules/head.py#L21-L22

**Comment by DonyMimin** [comment]
![image](https://github.com/user-attachments/assets/39fbddfb-9b04-40ba-b419-f0f4b780e45e)
thanks for the answer! but i'm still not understand on how it predict the class probabilities, if you mind please brief a bit the algorithm of the model YOLOv8. i appreciate it a lot for your answer

**Comment by glenn-jocher** [comment]
YOLOv8 predicts class probabilities using a classification head that processes features from the backbone and neck, separate from the bounding box predictions. This involves convolutional layers that output class scores for each detected object. For more details, you can explore the YOLOv8 architecture in the documentation.

**Comment by DonyMimin** [comment]
Thank you very much for the answer @glenn-jocher, it was really helpful

**Comment by DonyMimin** [comment]
> @DonyMimin no, YOLOv8 uses only convolutional layers in the head and does not use fully connected or "linear" layers, other than in the Classify models. Detect() heads are here:
> 
> https://github.com/ultralytics/ultralytics/blob/c5bee7545cb5fe26e201c593f3920b4b462c73c1/ultralytics/nn/modules/head.py#L21-L22

According to this information, just want to clarify it, so i found out that the code are using linear on classify, is this code are used to predict classes? (ultralytics/nn/modules/head.py#L252-L282)

    def __init__(self, c1, c2, k=1, s=1, p=None, g=1):
        """Initializes YOLOv8 classification head to transform input tensor from (b,c1,20,20) to (b,c2) shape."""
        super().__init__()
        c_ = 1280  # efficientnet_b0 size
        self.conv = Conv(c1, c_, k, s, p, g)
        self.pool = nn.AdaptiveAvgPool2d(1)  # to x(b,c_,1,1)
        self.drop = nn.Dropout(p=0.0, inplace=True)
        self.linear = nn.Linear(c_, c2)  # to x(b,c2)

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions about YOLOv8, feel free to ask. The community and documentation are great resources for further exploration.

**Comment by DonyMimin** [comment]
> YOLOv8 predicts class probabilities using a classification head that processes features from the backbone and neck, separate from the bounding box predictions. This involves convolutional layers that output class scores for each detected object. For more details, you can explore the YOLOv8 architecture in the documentation.

hi @glenn-jocher , i'm still curious about the convolutional layers that output class scores for each detected object. does it predict the classes based on the information of the feature-maps?

**Comment by DonyMimin** [comment]
hi @glenn-jocher , just for clarify, yolov8 is not using any fully connected right? I appreciate for your answer

**Comment by DonyMimin** [comment]
> YOLOv8 predicts class probabilities using a classification head that processes features from the backbone and neck, separate from the bounding box predictions. This involves convolutional layers that output class scores for each detected object. For more details, you can explore the YOLOv8 architecture in the documentation.

about this information, i want to ask how it works the "classification head" to know the class of the object? i really appreciate for your fast answer

**Comment by glenn-jocher** [comment]
The classification head in YOLOv8 uses convolutional layers to process features and predict class probabilities. It learns to recognize patterns associated with each class during training. This is achieved through a single neural network that optimizes both detection and classification tasks simultaneously. If you need more details, the documentation provides a deeper dive into the architecture.

**Comment by DonyMimin** [comment]
Thanks for the answer, really appreciate it

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.
